File: 2019/App/Day1.cs
using System.Linq;

namespace App
{
    public class Day1 : Day
    {
        protected override string Part1Code(string data) => data
                .Lines()
                .Select(int.Parse)
                .Select(FuelForMass)
                .Sum()
                .ToString();

        protected override string Part2Code(string data) => data
                .Lines()
                .Select(int.Parse)
                .Select(AddFuelRecursively)
                .Sum()
                .ToString();

        private static int FuelForMass(int mass) => (mass / 3) - 2;

        private static int AddFuelRecursively(int mass)
        {
            var massOfAddFuel = FuelForMass(mass);
            return massOfAddFuel <= 0 ? 0 : massOfAddFuel + AddFuelRecursively(massOfAddFuel);
        }
    }
}

File: 2019/js/day1.js
var fs = require('fs');
var path = require('path');
var filePath = path.join(__dirname, 'day1.txt');
var contents = fs.readFileSync(filePath, 'utf8');
var lines = contents.split('\n');

var fuelForMass = mass => (mass/3|0)-2;

// part 1
var sum = lines
    .map(line => fuelForMass(parseInt(line)))
    .reduce((p,f) => p+f, 0);
console.log(sum);

var fuelForMassRecursively = mass =>
{
    var massOfAddFuel = fuelForMass(mass);
    return massOfAddFuel <= 0 ? 0 : massOfAddFuel + fuelForMassRecursively(massOfAddFuel);
}
// part 2
sum = lines
    .map(line => fuelForMassRecursively(parseInt(line)))
    .reduce((p,f) => p+f, 0);
console.log(sum);

